File: fe/src/utils/shareCode.ts
const CLOUD189_SHARE_LINK_RE =
  /(?:^|[^a-z0-9_+./:@-])(?:(?:https?:\/\/|\/\/)?)(?:www\.)?cloud\.189\.cn(?::\d+)?\/t\/([^\s/?#()（）,，;；。:：]+)/i
const CLOUD189_SHARE_CODE_RE =
  /(?:分享码|\bshare[\s_-]?code\b|\bshare\b)[:：]\s*([^\s,，;；。()（）]+)/i
const CLOUD189_ACCESS_CODE_RE =
  /(访问码|提取码|\baccess[\s_-]?code\b|\baccess\b|\bcode\b)[:：]\s*([^\s,，;；。()（）]+)/gi
const CLOUD189_LABEL_PREFIX_RE =
  /^(?:访问码|提取码|分享码|\baccess[\s_-]?code\b|\baccess\b|\bcode\b|\bshare[\s_-]?code\b|\bshare\b)[:：]?/i
const CLOUD189_BOUNDARY_TRIM_RE =
  /^[\s.,，。;；:：!?！？、()[\]【】<>《》「」『』“”‘’"']+|[\s.,，。;；:：!?！？、()[\]【】<>《》「」『』“”‘’"']+$/g
const CLOUD189_EMBEDDED_LABEL_SEPARATOR_RE = /[、!！?？\]】》〉」』”’]/gu
const CLOUD189_RAW_CODE_RE = /^[a-z0-9]+$/i
const CLOUD189_URL_LIKE_RE =
  /(?:[a-z][a-z0-9+.-]*:\/\/|\/\/)[^\s"'<>]+|(?:^|[^a-z0-9_+./:@-])(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,}(?::\d+)?(?:\/[^\s"'<>]*|\?[^\s"'<>]*)/i

export interface Cloud189ShareParams {
  shareCode: string
  accessCode: string
}

export const isCloud189ShareCode = (value: string) => {
  return CLOUD189_RAW_CODE_RE.test(value.trim())
}

export const isCloud189AccessCode = (value: string) => {
  return CLOUD189_RAW_CODE_RE.test(value.trim())
}

export const containsCloud189URLLike = (value: string) => {
  return CLOUD189_URL_LIKE_RE.test(value.trim())
}

export const parseCloud189ShareCode = (
  input: string,
  explicitAccessCode = ''
): Cloud189ShareParams => {
  const cleanCode = normalizeCloud189ShareText(input)
  const normalizedExplicitAccessCode = normalizeExplicitCloud189AccessCode(explicitAccessCode)

  return {
    shareCode: extractCloud189ShareCode(cleanCode),
    accessCode: normalizedExplicitAccessCode || extractCloud189AccessCode(cleanCode),
  }
}

const normalizeExplicitCloud189AccessCode = (input: string) => {
  const cleanCode = normalizeCloud189ShareText(input)
  if (!cleanCode) {
    return ''
  }

  const accessCode = extractCloud189AccessCode(cleanCode)
  if (accessCode) {
    return accessCode
  }

  return cleanCloud189ExtractedCodeCandidate(cleanCode)
}

const normalizeCloud189ShareText = (input: string) => {
  return input.replace(/（/g, '(').replace(/）/g, ')').replace(/：/g, ':').trim()
}

const extractCloud189AccessCode = (cleanCode: string) => {
  CLOUD189_ACCESS_CODE_RE.lastIndex = 0

  for (const match of cleanCode.matchAll(CLOUD189_ACCESS_CODE_RE)) {
    const label = match[1]?.toLowerCase() || ''
    const index = match.index ?? 0
    if (label === 'code' && hasShareLabelBeforeCode(cleanCode.slice(0, index))) {
      continue
    }

    return cleanCloud189ExtractedCodeCandidate(match[2] || '')
  }

  const parts = cleanCode.split(/\s+/).filter(Boolean)
  if (parts.length <= 1) {
    return ''
  }

  const lastPart = cleanCloud189ExtractedCodeCandidate(parts[parts.length - 1])
  if (lastPart.length === 4) {
    return lastPart
  }

  return ''
}

const hasShareLabelBeforeCode = (prefix: string) => {
  return prefix
    .trim()
    .toLowerCase()
    .replace(/[-_\s]+$/g, '')
    .endsWith('share')
}

const cleanCloud189ExtractedCodeCandidate = (value: string) => {
  const trimmedValue = value.trim()

  return trimCloud189CodeBoundary(cutCloud189EmbeddedLabel(trimmedValue))
}

const cutCloud189EmbeddedLabel = (value: string) => {
  for (const match of value.matchAll(CLOUD189_EMBEDDED_LABEL_SEPARATOR_RE)) {
    const index = match.index ?? -1
    if (index < 1) {
      continue
    }

    const prefix = value.slice(0, index).trim()
    const suffix = trimCloud189CodeBoundary(value.slice(index + match[0].length))
    if (prefix && CLOUD189_LABEL_PREFIX_RE.test(suffix)) {
      return prefix
    }
  }

  return value
}

const trimCloud189CodeBoundary = (value: string) => {
  return value.replace(CLOUD189_BOUNDARY_TRIM_RE, '')
}

const extractCloud189ShareCode = (cleanCode: string) => {
  const linkMatch = cleanCode.match(CLOUD189_SHARE_LINK_RE)
  if (linkMatch?.[1]) {
    return cleanCloud189ExtractedCodeCandidate(linkMatch[1])
  }

  const labelMatch = cleanCode.match(CLOUD189_SHARE_CODE_RE)
  if (labelMatch?.[1]) {
    return cleanCloud189ExtractedCodeCandidate(labelMatch[1])
  }

  const bracketIndex = cleanCode.indexOf('(')
  if (bracketIndex > -1) {
    return cleanCode.slice(0, bracketIndex).trim()
  }

  if (CLOUD189_URL_LIKE_RE.test(cleanCode)) {
    return ''
  }

  const parts = cleanCode.split(/\s+/).filter(Boolean)
  if (parts.length > 0) {
    return cleanCloud189ExtractedCodeCandidate(parts[0])
  }

  return cleanCode
}

File: internal/pkgs/utils/share_code.go
package utils

import (
	"regexp"
	"strings"
)

var (
	cloud189ShareLinkPattern  = regexp.MustCompile(`(?i)(?:^|[^a-z0-9_+./:@-])(?:(?:https?:\/\/|\/\/)?)(?:www\.)?cloud\.189\.cn(?::\d+)?\/t\/([^\s/?#()（）,，;；。:：]+)`)
	cloud189ShareCodePattern  = regexp.MustCompile(`(?i)(?:分享码|\bshare[\s_-]?code\b|\bshare\b)[:：]\s*([^\s,，;；。()（）]+)`)
	cloud189AccessCodePattern = regexp.MustCompile(`(?i)(访问码|提取码|\baccess[\s_-]?code\b|\baccess\b|\bcode\b)[:：]\s*([^\s,，;；。()（）]+)`)
	cloud189LabelPrefix       = regexp.MustCompile(`(?i)^(?:访问码|提取码|分享码|\baccess[\s_-]?code\b|\baccess\b|\bcode\b|\bshare[\s_-]?code\b|\bshare\b)[:：]?`)
	cloud189ShareCodeValue    = regexp.MustCompile(`^[a-zA-Z0-9]+$`)
	cloud189AccessCodeValue   = regexp.MustCompile(`^[a-zA-Z0-9]+$`)
	cloud189URLLikePattern    = regexp.MustCompile(`(?i)(?:[a-z][a-z0-9+.-]*://|//)[^\s"'<>]+|(?:^|[^a-z0-9_+./:@-])(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,}(?::\d+)?(?:/[^\s"'<>]*|\?[^\s"'<>]*)`)
)

const (
	cloud189CodeBoundaryTrimChars       = " \t\r\n.,，。;；:：!?！？、()[]【】<>《》「」『』“”‘’\"'"
	cloud189EmbeddedLabelSeparatorChars = "、!！?？]】》〉」』”’"
)

func ParseCloud189ShareCode(input string, explicitAccessCode string) (string, string) {
	cleanCode := normalizeCloud189ShareCode(input)
	accessCode := extractCloud189AccessCode(cleanCode)

	explicitAccessCode = normalizeExplicitCloud189AccessCode(explicitAccessCode)
	if explicitAccessCode != "" {
		accessCode = explicitAccessCode
	}

	return extractCloud189ShareCode(cleanCode), accessCode
}

func IsCloud189ShareLink(value string) bool {
	return cloud189ShareLinkPattern.MatchString(normalizeCloud189ShareCode(value))
}

// ContainsURLLike 判断文本中是否包含 URL 形态的片段。
func ContainsURLLike(value string) bool {
	return cloud189URLLikePattern.MatchString(strings.TrimSpace(value))
}

func IsCloud189ShareCode(value string) bool {
	value = strings.TrimSpace(value)
	if value == "" {
		return false
	}

	return cloud189ShareCodeValue.MatchString(value)
}

func IsCloud189AccessCode(value string) bool {
	value = strings.TrimSpace(value)
	if value == "" {
		return false
	}

	return cloud189AccessCodeValue.MatchString(value)
}

func normalizeExplicitCloud189AccessCode(input string) string {
	cleanCode := normalizeCloud189ShareCode(input)
	if cleanCode == "" {
		return ""
	}

	if accessCode := extractCloud189AccessCode(cleanCode); accessCode != "" {
		return accessCode
	}

	return cleanCloud189ExtractedCodeCandidate(cleanCode)
}

func normalizeCloud189ShareCode(input string) string {
	cleanCode := strings.ReplaceAll(input, "（", "(")
	cleanCode = strings.ReplaceAll(cleanCode, "）", ")")
	cleanCode = strings.ReplaceAll(cleanCode, "：", ":")

	return strings.TrimSpace(cleanCode)
}

func extractCloud189AccessCode(cleanCode string) string {
	codeMatches := cloud189AccessCodePattern.FindAllStringSubmatchIndex(cleanCode, -1)
	for _, codeMatch := range codeMatches {
		if len(codeMatch) < 6 {
			continue
		}

		label := strings.ToLower(cleanCode[codeMatch[2]:codeMatch[3]])
		if label == "code" && hasShareLabelBeforeCode(cleanCode[:codeMatch[0]]) {
			continue
		}

		return cleanCloud189ExtractedCodeCandidate(cleanCode[codeMatch[4]:codeMatch[5]])
	}

	parts := strings.Fields(cleanCode)
	if len(parts) <= 1 {
		return ""
	}

	lastPart := cleanCloud189ExtractedCodeCandidate(parts[len(parts)-1])
	if len(lastPart) == 4 {
		return lastPart
	}

	return ""
}

func hasShareLabelBeforeCode(prefix string) bool {
	prefix = strings.ToLower(strings.TrimSpace(prefix))
	prefix = strings.TrimRight(prefix, "_- ")

	return strings.HasSuffix(prefix, "share")
}

func cleanCloud189ExtractedCodeCandidate(value string) string {
	value = strings.TrimSpace(value)
	value = cutCloud189EmbeddedLabel(value)

	return strings.Trim(value, cloud189CodeBoundaryTrimChars)
}

func cutCloud189EmbeddedLabel(value string) string {
	for idx, char := range value {
		if !strings.ContainsRune(cloud189EmbeddedLabelSeparatorChars, char) {
			continue
		}

		prefix := strings.TrimSpace(value[:idx])

		suffix := strings.Trim(value[idx+len(string(char)):], cloud189CodeBoundaryTrimChars)
		if prefix != "" && cloud189LabelPrefix.MatchString(suffix) {
			return prefix
		}
	}

	return value
}

func extractCloud189ShareCode(cleanCode string) string {
	if matches := cloud189ShareLinkPattern.FindStringSubmatch(cleanCode); len(matches) > 1 {
		return cleanCloud189ExtractedCodeCandidate(matches[1])
	}

	if matches := cloud189ShareCodePattern.FindStringSubmatch(cleanCode); len(matches) > 1 {
		return cleanCloud189ExtractedCodeCandidate(matches[1])
	}

	if idx := strings.Index(cleanCode, "("); idx > -1 {
		return strings.TrimSpace(cleanCode[:idx])
	}

	if ContainsURLLike(cleanCode) {
		return ""
	}

	parts := strings.Fields(cleanCode)
	if len(parts) > 0 {
		return cleanCloud189ExtractedCodeCandidate(parts[0])
	}

	return cleanCode
}
